File: python/dr_companion.py
"""The Python side of the scripting API decided in docs/ENGINE.md.

Talks to the app over the loopback socket `src-tauri/src/script_api.rs`
starts on every launch: newline-delimited JSON, authenticated by a token the
app writes beside its own data on startup. This library exists so nobody
writing a script has to know that - it finds the port and token, connects,
and turns each incoming JSON line into a callback.

Pure standard library, on purpose. `pip install` is a step between a script
idea and running it, and the whole reason a scripting layer exists is to
lower that distance. `python your_script.py` should be the entire setup.

# What a line looks like right now, and the gap that is not hidden

A script receives the same wire text this app's own frontend receives before
its TypeScript parser (`src/lib/gameStream.ts`) turns it into clean lines with
a channel and a bold flag. That parser is hardened by several rounds of
finding real bugs in it - a tag split across two reads, a literal `<` in game
text capturing a real `</popStream>` - and porting it to Python untested
against the same cases would risk a second, quietly different parser rather
than closing a known gap. So `Line.text` can contain markup like
`<pushStream id='thoughts'/>` today. See docs/PYTHON_API.md for the fuller
version of this note and where the gap is tracked.

# Example

    from dr_companion import Companion

    c = Companion()

    @c.on_line
    def watch(line):
        if "you are stunned" in line.text.lower():
            c.send("stand")

    c.run()
"""

from __future__ import annotations

import json
import os
import socket
import threading
import time
from dataclasses import dataclass
from pathlib import Path
from typing import Callable, Optional


class ConnectionError(Exception):
    """Could not reach the app, or it refused the token."""


class NotConnected(Exception):
    """A method that needs a live socket was called without one."""


@dataclass(frozen=True)
class Line:
    """One chunk of game text. `seq` is stable and increasing; `text` is not
    guaranteed to be one visual line - see the module note on markup."""

    seq: int
    text: str


@dataclass(frozen=True)
class Status:
    """The app's own connection to Lich, as of the last time it was asked or
    the last time it changed - not necessarily this instant. `connected`
    turning `False` mid-script means Lich hung up; it does not mean this
    library disconnected the script."""

    connected: bool
    host: str
    port: int
    lines: int
    note: str


def _data_dir() -> Path:
    """Where the app keeps `script-api.token` and `script-api.port`.

    Mirrors `setup::app_data_dir()` in the Rust side exactly - `%LOCALAPPDATA%`
    joined with the same folder name. If that environment variable is ever
    absent (it is not, on any Windows this app supports), this raises rather
    than guessing a folder the app never wrote to, which would produce a
    confusing "token not found" instead of an honest "cannot locate app data".
    """
    local_app_data = os.environ.get("LOCALAPPDATA")
    if not local_app_data:
        raise ConnectionError(
            "LOCALAPPDATA is not set - cannot find where DR Companion keeps its data. "
            "Pass host/port/token to Companion() explicitly instead."
        )
    return Path(local_app_data) / "DR Companion Data"


def _read_connection_info(data_dir: Path) -> tuple[int, str]:
    port_file = data_dir / "script-api.port"
    token_file = data_dir / "script-api.token"

    if not port_file.exists() or not token_file.exists():
        raise ConnectionError(
            f"No script API files found in {data_dir}. Is DR Companion running? "
            "It writes these on startup."
        )

    port_text = port_file.read_text(encoding="utf-8").strip()
    if not port_text.isdigit():
        raise ConnectionError(f"{port_file} does not contain a port number: {port_text!r}")

    token = token_file.read_text(encoding="utf-8").strip()
    if not token:
        raise ConnectionError(f"{token_file} is empty")

    return int(port_text), token


class Companion:
    """One connection to the app, and the callbacks a script hangs off it.

    Not thread-safe to call `send()` from inside an `on_line` callback and
    also from another thread without your own lock - the underlying socket
    write is not synchronised, on the theory that a script is single-threaded
    unless it deliberately isn't, and adding a lock for a case that mostly
    does not happen would tax every script for the rare one that needs it.
    """

    def __init__(
        self,
        host: str = "127.0.0.1",
        port: Optional[int] = None,
        token: Optional[str] = None,
        connect_timeout: float = 5.0,
    ) -> None:
        if port is None or token is None:
            found_port, found_token = _read_connection_info(_data_dir())
            port = port if port is not None else found_port
            token = token if token is not None else found_token

        self._host = host
        self._port = port
        self._token = token
        self._sock: Optional[socket.socket] = None
        self._buf = b""
        self._line_handlers: list[Callable[[Line], None]] = []
        self._state_handlers: list[Callable[[Status], None]] = []
        self._connect_timeout = connect_timeout
        self._stopped = threading.Event()

    # -- connecting -----------------------------------------------------

    def connect(self) -> None:
        """Open the socket and authenticate. Raises `ConnectionError` on
        anything short of success - there is no partially-connected state a
        caller needs to check for."""
        try:
            sock = socket.create_connection(
                (self._host, self._port), timeout=self._connect_timeout
            )
        except OSError as e:
            # A stale port file and a live one are the same file.
            #
            # The port and token are rewritten every time DR Companion starts,
            # and the files survive it closing - so pointing at a port nothing
            # is listening on is the *normal* failure here, not an exotic one.
            # Unhandled, it surfaced as `ConnectionRefusedError: [WinError
            # 10061] No connection could be made because the target machine
            # actively refused it`, ten frames deep, naming a port number the
            # reader never chose and saying nothing about the app.
            #
            # Hit while writing the first task on top of this library: the app
            # had been closed, the file still held its last port, and the error
            # gave no hint that "start DR Companion" was the whole fix.
            raise ConnectionError(
                f"Nothing is listening on {self._host}:{self._port}.\n"
                f"  That port came from {_data_dir() / 'script-api.port'}, which DR "
                "Companion rewrites every time it starts and leaves behind when it "
                "closes.\n"
                "  So this almost always means the app is not running. Start DR "
                "Companion and try again.\n"
                f"  (underlying error: {e})"
            ) from e

        sock.settimeout(self._connect_timeout)
        self._sock = sock

        hello = self._read_message()
        if hello is None or hello.get("type") != "hello":
            self.close()
            raise ConnectionError(f"expected a hello frame, got {hello!r}")

        self._write_message({"type": "auth", "token": self._token})
        reply = self._read_message()
        if reply is None or reply.get("type") != "auth_ok":
            self.close()
            raise ConnectionError(
                "the app refused this token. If DR Companion was restarted, "
                "its token changed - reread it rather than reusing an old one."
            )

        # No more timeout once authenticated: a script legitimately waits
        # quietly for a long time, and a socket timeout would turn "nothing
        # happened yet" into a spurious exception.
        sock.settimeout(None)

    def close(self) -> None:
        self._stopped.set()
        if self._sock is not None:
            try:
                self._sock.close()
            except OSError:
                pass
            self._sock = None

    # -- sending ----------------------------------------------------------

    def send(self, command: str) -> None:
        """Send a command exactly as typed - no aliasing, no interpretation,
        matching the Rust side's own `game_send`."""
        if self._sock is None:
            self.connect()
        self._write_message({"type": "send", "command": command})

    def status(self) -> Status:
        """Ask once, rather than waiting for the next state broadcast.
        Blocks briefly for the reply; a script that wants continuous updates
        should use `on_state` instead.

        Connects first if `connect()` was not already called - found by
        actually running the example script rather than by inspection: it
        called `status()` before `run()` had a chance to connect, and got a
        `NotConnected` that a docstring cannot fix. `run()` already tolerated
        this; `send()` and `status()` now match it, so a script can call any
        of the three first without needing to know one of them is special.
        """
        if self._sock is None:
            self.connect()
        self._write_message({"type": "status"})
        while True:
            msg = self._read_message()
            if msg is None:
                raise ConnectionError("connection closed while waiting for status")
            if msg.get("type") == "state":
                return _status_from(msg)
            # Anything else (e.g. a line that arrived first) goes through the
            # normal dispatch so it is not silently dropped just because this
            # call was waiting for something else.
            self._dispatch(msg)

    # -- callbacks ----------------------------------------------------------

    def on_line(self, fn: Callable[[Line], None]) -> Callable[[Line], None]:
        """Usable as a decorator or a plain call. Multiple handlers may be
        registered; each sees every line, in registration order."""
        self._line_handlers.append(fn)
        return fn

    def on_state(self, fn: Callable[[Status], None]) -> Callable[[Status], None]:
        self._state_handlers.append(fn)
        return fn

    # -- the loop ----------------------------------------------------------

    def run(self) -> None:
        """Connect if not already connected, then read forever, dispatching
        to callbacks, until the connection closes or `stop()` is called from
        another thread (e.g. a handler running in a background thread, or a
        signal handler)."""
        if self._sock is None:
            self.connect()

        while not self._stopped.is_set():
            msg = self._read_message()
            if msg is None:
                break
            self._dispatch(msg)

    def stop(self) -> None:
        self._stopped.set()

    def _dispatch(self, msg: dict) -> None:
        kind = msg.get("type")
        if kind == "line":
            line = Line(seq=msg.get("seq", 0), text=msg.get("text", ""))
            for fn in self._line_handlers:
                fn(line)
        elif kind == "state":
            status = _status_from(msg)
            for fn in self._state_handlers:
                fn(status)
        elif kind == "error":
            # Surfaced rather than swallowed. A script that sends a malformed
            # request deserves to see why nothing happened, not a silent gap.
            print(f"dr_companion: the app reported an error: {msg.get('message')}")
        # Unknown message types are ignored rather than raising, so a future
        # message kind this version of the library does not know about does
        # not crash an already-running script.

    # -- wire framing --------------------------------------------------

    def _write_message(self, obj: dict) -> None:
        if self._sock is None:
            raise NotConnected("call connect() first")
        line = json.dumps(obj) + "\n"
        self._sock.sendall(line.encode("utf-8"))

    def _read_message(self) -> Optional[dict]:
        """One JSON object, or `None` on a clean close.

        Buffers across `recv()` calls because a message and the start of the
        next one can arrive in the same packet, or one message can be split
        across two - the same reason `game_link.rs` reads bytes rather than
        trusting a socket to hand back exactly one line at a time.
        """
        if self._sock is None:
            raise NotConnected("call connect() first")

        while b"\n" not in self._buf:
            chunk = self._sock.recv(4096)
            if not chunk:
                return None
            self._buf += chunk

        line, self._buf = self._buf.split(b"\n", 1)
        text = line.decode("utf-8", errors="replace").strip()
        if not text:
            return self._read_message()
        try:
            return json.loads(text)
        except json.JSONDecodeError:
            # The app is trusted to send well-formed JSON; if this ever
            # fires, something upstream broke and the honest response is to
            # say so rather than skip the line and continue on bad data.
            raise ConnectionError(f"could not parse a message from the app: {text!r}")


def _status_from(msg: dict) -> Status:
    return Status(
        connected=bool(msg.get("connected", False)),
        host=msg.get("host", ""),
        port=int(msg.get("port", 0)),
        lines=int(msg.get("lines", 0)),
        note=msg.get("note", ""),
    )

File: typescript/dr_companion.ts
/**
 * The TypeScript side of the scripting API decided in docs/ENGINE.md.
 *
 * Same transport `python/dr_companion.py` talks to - `src-tauri/src/script_api.rs`'s
 * loopback socket, newline-delimited JSON, authenticated by a token the app
 * writes beside its own data on startup. That server was already
 * language-agnostic (`python/dr_companion.py`'s own docs say as much: "If you
 * are not using dr_companion.py - a script in another language, say - this is
 * everything it does for you"), so this is not new server-side work, only a
 * second client for a protocol that was never Python-specific.
 *
 * Zero npm dependencies, on purpose, for the same reason `dr_companion.py` is
 * pure standard library: `npm install` is a step between a script idea and
 * running it, and Node's own `net`/`fs`/`events` modules are enough. On
 * Node 24 (this project's stated minimum, see DEPENDENCIES.md), a `.ts` file
 * runs directly - `node your_script.ts` - no build step, the same "just run
 * it" promise Python makes. Older Node 22.6+ needs
 * `--experimental-strip-types` for the same thing; verified in this sandbox
 * at Node 22.22 with that flag.
 *
 * **The one real constraint that comes with "no build step":** Node's type
 * stripping erases type annotations only - it does not transpile TypeScript
 * features that generate actual code. Constructor parameter properties
 * (`constructor(public x: number) {}`) are the one `drtask.ts` hit first
 * (`SyntaxError [ERR_UNSUPPORTED_TYPESCRIPT_SYNTAX]`); enums with values and
 * `namespace` blocks are the same category of problem. Assign fields the
 * plain way in a script's own code and this never comes up.
 *
 * # Where this deliberately differs from the Python client, and why
 *
 * `dr_companion.py`'s `Companion` blocks: `run()` loops on a synchronous
 * socket read and calls your handlers in between. Node has no equivalent
 * "blocking read" without spinning up a worker thread to fake it, and faking
 * it would fight the platform instead of using it. So this `Companion` is an
 * ordinary Node `EventEmitter` over an async socket: `on('line', ...)` /
 * `on('state', ...)` instead of `on_line`/`on_state`, and the process stays
 * alive because the socket has listeners, the same as any other Node network
 * client - there is no `run()` loop to call. `connect()`, `send()` and
 * `status()` are `async` because the underlying I/O is.
 *
 * # The same known gap Python has
 *
 * `Line.text` is the same raw wire chunk the Python client receives - see
 * `dr_companion.py`'s module docstring for the fuller version of this note.
 * `<pushStream id='thoughts'/>` markup is not parsed out here either.
 *
 * # Example
 *
 *     import { Companion } from './dr_companion.ts'
 *
 *     const c = new Companion()
 *     c.on('line', (line) => {
 *       if (line.text.toLowerCase().includes('you are stunned')) {
 *         c.send('stand')
 *       }
 *     })
 *     await c.connect()
 */

import { EventEmitter } from 'node:events'
import { existsSync, readFileSync } from 'node:fs'
import { createConnection, type Socket } from 'node:net'
import * as path from 'node:path'

export class ConnectionError extends Error {}
export class NotConnected extends Error {}

/** One chunk of game text. `seq` is stable and increasing; `text` is not
 * guaranteed to be one visual line - see the module note on markup. */
export interface Line {
  seq: number
  text: string
}

/** The app's own connection to Lich, as of the last time it was asked or the
 * last time it changed - not necessarily this instant. */
export interface Status {
  connected: boolean
  host: string
  port: number
  lines: number
  note: string
}

interface ConnectionInfo {
  port: number
  token: string
}

/** Mirrors `setup::app_data_dir()` on the Rust side - `%LOCALAPPDATA%` joined
 * with the app's data folder name. Throws rather than guessing a folder the
 * app never wrote to if the environment variable is absent. */
function dataDir(): string {
  const localAppData = process.env.LOCALAPPDATA
  if (!localAppData) {
    throw new ConnectionError(
      'LOCALAPPDATA is not set - cannot find where DR Companion keeps its data. ' +
        'Pass host/port/token to Companion() explicitly instead.',
    )
  }
  return path.join(localAppData, 'DR Companion Data')
}

function readConnectionInfo(dir: string): ConnectionInfo {
  const portFile = path.join(dir, 'script-api.port')
  const tokenFile = path.join(dir, 'script-api.token')

  if (!existsSync(portFile) || !existsSync(tokenFile)) {
    throw new ConnectionError(
      `No script API files found in ${dir}. Is DR Companion running? It writes these on startup.`,
    )
  }

  const portText = readFileSync(portFile, 'utf8').trim()
  const port = Number(portText)
  if (!Number.isInteger(port)) {
    throw new ConnectionError(`${portFile} does not contain a port number: ${JSON.stringify(portText)}`)
  }

  const token = readFileSync(tokenFile, 'utf8').trim()
  if (!token) {
    throw new ConnectionError(`${tokenFile} is empty`)
  }

  return { port, token }
}

function statusFrom(msg: Record<string, unknown>): Status {
  return {
    connected: Boolean(msg.connected),
    host: typeof msg.host === 'string' ? msg.host : '',
    port: typeof msg.port === 'number' ? msg.port : 0,
    lines: typeof msg.lines === 'number' ? msg.lines : 0,
    note: typeof msg.note === 'string' ? msg.note : '',
  }
}

export interface CompanionOptions {
  host?: string
  port?: number
  token?: string
  connectTimeoutMs?: number
}

export interface Companion {
  on(event: 'line', listener: (line: Line) => void): this
  on(event: 'state', listener: (status: Status) => void): this
  on(event: 'error', listener: (message: string) => void): this
  on(event: 'close', listener: () => void): this
}

/** One connection to the app, and the events a script listens for on it. */
export class Companion extends EventEmitter {
  private readonly host: string
  private port: number | undefined
  private token: string | undefined
  private readonly connectTimeoutMs: number
  private socket: Socket | null = null
  private buf = ''
  private statusWaiters: Array<(s: Status) => void> = []

  constructor(opts: CompanionOptions = {}) {
    super()
    if (opts.port === undefined || opts.token === undefined) {
      const found = readConnectionInfo(dataDir())
      opts.port = opts.port ?? found.port
      opts.token = opts.token ?? found.token
    }
    this.host = opts.host ?? '127.0.0.1'
    this.port = opts.port
    this.token = opts.token
    this.connectTimeoutMs = opts.connectTimeoutMs ?? 5000
  }

  /** Opens the socket and authenticates. Rejects on anything short of
   * success - there is no partially-connected state a caller needs to check
   * for. */
  async connect(): Promise<void> {
    if (this.port === undefined || this.token === undefined) {
      throw new ConnectionError('no port/token available')
    }
    const port = this.port
    const token = this.token

    const socket = await new Promise<Socket>((resolve, reject) => {
      const s = createConnection({ host: this.host, port }, () => resolve(s))
      s.setTimeout(this.connectTimeoutMs)
      s.once('timeout', () => reject(new ConnectionError(`connect to ${this.host}:${port} timed out`)))
      s.once('error', (e) => reject(new ConnectionError(String(e))))
    })
    this.socket = socket
    this.buf = ''

    const hello = await this.readOneMessage()
    if (hello === null || hello.type !== 'hello') {
      this.close()
      throw new ConnectionError(`expected a hello frame, got ${JSON.stringify(hello)}`)
    }

    this.writeMessage({ type: 'auth', token })
    const reply = await this.readOneMessage()
    if (reply === null || reply.type !== 'auth_ok') {
      this.close()
      throw new ConnectionError(
        'the app refused this token. If DR Companion was restarted, its token changed - reread it rather than reusing an old one.',
      )
    }

    socket.setTimeout(0)
    socket.on('data', (chunk) => this.onData(chunk))
    socket.on('close', () => this.emit('close'))
    socket.on('error', (e) => this.emit('error', String(e)))
  }

  close(): void {
    if (this.socket) {
      this.socket.destroy()
      this.socket = null
    }
  }

  /** Sends a command exactly as typed - no aliasing, no interpretation,
   * matching the Rust side's own `game_send`. Connects first if not already
   * connected. */
  async send(command: string): Promise<void> {
    if (this.socket === null) await this.connect()
    this.writeMessage({ type: 'send', command })
  }

  /** Asks the app once for its state, rather than waiting for the next
   * unprompted broadcast. Any `line` (or other) message that arrives first
   * is still emitted normally through the usual events - waiting for status
   * never swallows it. */
  async status(): Promise<Status> {
    if (this.socket === null) await this.connect()
    return new Promise<Status>((resolve) => {
      this.statusWaiters.push(resolve)
      this.writeMessage({ type: 'status' })
    })
  }

  private writeMessage(obj: Record<string, unknown>): void {
    if (this.socket === null) throw new NotConnected('call connect() first')
    this.socket.write(JSON.stringify(obj) + '\n')
  }

  /** One JSON object read directly off the socket, or `null` on a clean
   * close - used only for the handshake, before `data` events take over. */
  private readOneMessage(): Promise<Record<string, unknown> | null> {
    if (this.socket === null) throw new NotConnected('call connect() first')
    const socket = this.socket
    return new Promise((resolve, reject) => {
      const onData = (chunk: Buffer) => {
        this.buf += chunk.toString('utf8')
        const nl = this.buf.indexOf('\n')
        if (nl === -1) return
        const line = this.buf.slice(0, nl)
        this.buf = this.buf.slice(nl + 1)
        cleanup()
        if (!line.trim()) return resolve(null)
        try {
          resolve(JSON.parse(line))
        } catch {
          reject(new ConnectionError(`could not parse a message from the app: ${JSON.stringify(line)}`))
        }
      }
      const onClose = () => {
        cleanup()
        resolve(null)
      }
      const cleanup = () => {
        socket.off('data', onData)
        socket.off('close', onClose)
      }
      socket.on('data', onData)
      socket.on('close', onClose)
    })
  }

  /** Splits accumulated bytes on `\n` - a message and the start of the next
   * one can arrive in the same packet, or one message can be split across
   * two, same reasoning as `dr_companion.py`'s own `_read_message`. */
  private onData(chunk: Buffer): void {
    this.buf += chunk.toString('utf8')
    let nl: number
    while ((nl = this.buf.indexOf('\n')) !== -1) {
      const raw = this.buf.slice(0, nl)
      this.buf = this.buf.slice(nl + 1)
      if (!raw.trim()) continue
      let msg: Record<string, unknown>
      try {
        msg = JSON.parse(raw)
      } catch {
        this.emit('error', `could not parse a message from the app: ${JSON.stringify(raw)}`)
        continue
      }
      this.dispatch(msg)
    }
  }

  private dispatch(msg: Record<string, unknown>): void {
    switch (msg.type) {
      case 'line':
        this.emit('line', {
          seq: typeof msg.seq === 'number' ? msg.seq : 0,
          text: typeof msg.text === 'string' ? msg.text : '',
        } satisfies Line)
        break
      case 'state': {
        const status = statusFrom(msg)
        this.emit('state', status)
        const waiters = this.statusWaiters
        this.statusWaiters = []
        for (const resolve of waiters) resolve(status)
        break
      }
      case 'error':
        this.emit('error', typeof msg.message === 'string' ? msg.message : String(msg.message))
        break
      // Unknown message types are ignored, same reasoning as the Python
      // client: a future message kind this version doesn't know about
      // should not crash an already-running script.
    }
  }
}
